dot/parachain/collator-protocol/validator_side.go

```diff
@@ -0,0 +1,307 @@
+
+	fetchedCollations []parachaintypes.Collation
+	// track all active collators and their data
+	peerData map[peer.ID]PeerData
+
+	ourView View
+
+	// Keep track of all pending candidate collations
+	pendingCandidates map[common.Hash]CollationEvent
+}
+
+func (cpvs CollatorProtocolValidatorSide) getPeerIDFromCollatorID(collatorID parachaintypes.CollatorID,
+) (peer.ID, bool) {
+	for peerID, peerData := range cpvs.peerData {
+		if peerData.state.PeerState == Collating && peerData.state.CollatingPeerState.CollatorID == collatorID {
+			return peerID, true
+		}
+	}
+
+	return "", false
+}
+
+type CollateOn parachaintypes.ParaID
+
+type DistributeCollation struct {
+	CandidateReceipt parachaintypes.CandidateReceipt
+	PoV              parachaintypes.PoV
+}
+
+type ReportCollator parachaintypes.CollatorID
+
+type NetworkBridgeUpdate struct {
+	// TODO: not quite sure if we would need this or something similar to this
+}
+
+type SecondedOverseerMsg struct {
+	Parent common.Hash
+	Stmt   parachaintypes.StatementVDT
+}
+
+type Backed struct {
+	ParaID parachaintypes.ParaID
+	// Hash of the para head generated by candidate
+	ParaHead common.Hash
+}
+
+type InvalidOverseeMsg struct {
+	Parent           common.Hash
+	CandidateReceipt parachaintypes.CandidateReceipt
+}
+
+func (cpvs CollatorProtocolValidatorSide) processMessage(msg interface{}) error {
+	// run this function as a goroutine, ideally
+
+	switch msg := msg.(type) {
+	case CollateOn:
+		return fmt.Errorf("CollateOn %w", ErrNotExpectedOnValidatorSide)
+	case DistributeCollation:
+		return fmt.Errorf("DistributeCollation %w", ErrNotExpectedOnValidatorSide)
+	case ReportCollator:
+		peerID, ok := cpvs.getPeerIDFromCollatorID(parachaintypes.CollatorID(msg))
+		if !ok {
+			return ErrPeerIDNotFoundForCollator
+		}
+		cpvs.net.ReportPeer(peerset.ReputationChange{
+			Value:  peerset.ReportBadCollatorValue,
+			Reason: peerset.ReportBadCollatorReason,
+		}, peerID)
+	case NetworkBridgeUpdate:
+		// TODO: handle network message https://github.com/ChainSafe/gossamer/issues/3515
+		// https://github.com/paritytech/polkadot-sdk/blob/db3fd687262c68b115ab6724dfaa6a71d4a48a59/polkadot/node/network/collator-protocol/src/validator_side/mod.rs#L1457 //nolint
+	case SecondedOverseerMsg:
+		// TODO: handle seconded message https://github.com/ChainSafe/gossamer/issues/3516
+		// https://github.com/paritytech/polkadot-sdk/blob/db3fd687262c68b115ab6724dfaa6a71d4a48a59/polkadot/node/network/collator-protocol/src/validator_side/mod.rs#L1466 //nolint
+
+	case Backed:
+		// TODO: handle backed message https://github.com/ChainSafe/gossamer/issues/3517
+	case InvalidOverseeMsg:
+		invalidOverseerMsg := msg
+
+		collationEvent, ok := cpvs.pendingCandidates[invalidOverseerMsg.Parent]
+		if !ok {
+			return nil
+		}
+
+		if *collationEvent.PendingCollation.CommitmentHash == (invalidOverseerMsg.CandidateReceipt.CommitmentsHash) {
+			delete(cpvs.pendingCandidates, invalidOverseerMsg.Parent)
```

Comment:
should you introduce a `sync.Mutex` for protecting this map against race conditions?

Reply:
Added sync.Mutex.

I see multiple messages getting processes in async way. So, adding sync Mutex for pendingCandidates.

Reply:
Looks like I forgot to push the commit that added sync.Mutex. Will add it in the next PR.